lib/blocks.php

```diff
@@ -592,3 +592,108 @@ function gutenberg_multiple_block_styles( $metadata ) {
 	return $metadata;
 }
 add_filter( 'block_type_metadata', 'gutenberg_multiple_block_styles' );
+
+if ( ! function_exists( 'wp_enqueue_block_view_script' ) ) {
+	/**
+	 * Enqueue a frontend script for a specific block.
+	 *
+	 * Scripts enqueued using this function will only get printed
+	 * when the block gets rendered on the frontend.
+	 *
+	 * @param string $block_name The block-name, including namespace.
+	 * @param array  $args       An array of arguments [handle,src,deps,ver,media].
+	 *
+	 * @return void
+	 */
+	function wp_enqueue_block_view_script( $block_name, $args ) {
+		$args = wp_parse_args(
+			$args,
+			array(
+				'handle'     => '',
+				'src'        => '',
+				'deps'       => array(),
+				'ver'        => false,
+				'in_footer'  => false,
+
+				// Additional arg to allow translations for the script's textdomain.
+				'textdomain' => '',
```

Comment:
It's out of scope for this PR, but maybe we should make `textdomain` property of the block type? @swissspidy or @ocean90, what do you think?

Reply:
So a property of the `WP_Block_Type` class you mean? Isn't it part of `WP_Block_Type::$attributes` already?

Reply:
That's a great question. I didn't bother to check earlier because I assumed it isn't 🙈 

It surely is:

https://github.com/WordPress/wordpress-develop/blob/b72c1d00ae3a2b8fd6cd5f87837ae157d4231e55/src/wp-includes/class-wp-block-type.php#L86-L92

@aristath, that would be ideal to consume the textdomain directly from the block type.

Reply:
I'm a bit sceptical about this...
Suppose that I'm building a theme, and I want to add some additional behavior to the navigation menus.
I'll write a script in my theme, and then use the `wp_enqueue_block_view_script()` function to "attach" that script to the `core/navigation-link`, or `core/navigation-submenu` block depending on what my script does.
Chances are, the script that I've written will need to use the theme's textdomain and not the block's textdomain - which means I should be able to define my own textdomain for the script I'm enqueueing 🤔

Reply:
Right, it makes sense. The challange though would be if the theme uses non-standard folder for translations. You might need to offer a way to pass 3rd argument here:

```js
wp_set_script_translations( $args['handle'], $args['textdomain'] );
```

Reply:
I don't quite understand why we'd need to pass a 3rd arg to the `wp_set_script_translations` function... 🤔 
If that is the case, then shouldn't the `register_block_script_handle()` function do the same in core?
The assumption is that if they use a textdomain, they've used the `load_theme_textdomain` or `load_plugin_textdomain` functions to properly register the textdomain etc (or are using the default, standard folders for translations in which case these functions should not be necessary)
If a theme/plugin has working translations, then we shouldn't have to use the 3rd arg... Or am I missing something?

Reply:
If a theme/plugin uses custom translation folders they must use `load_plugin_textdomain`/`load_theme_textdomain`, correct. But that only covers _PHP_ translations. They still must pass the 3rd arg to `wp_set_script_translations` to cover _JS_ translations.

Not familiar with `register_block_script_handle` but yes at first glance it might need the same. Probably best to give it a try and see if it works :-)

Maybe a good opportunity to update https://github.com/swissspidy/gutenberg-i18n-block ...

Reply:
> If that is the case, then shouldn't the register_block_script_handle() function do the same in core?

Yes, it defintiely should. See: https://core.trac.wordpress.org/ticket/54797. I didn't have a chance to revisit the patch though.

Reply:
Well, we can leave it for later and tackle custom translation folders together.

Reply:
Added a separate PR for the `domainpath` on https://github.com/WordPress/gutenberg/pull/40136 so we can do it when the time is right, and ensure it's done the same way we handle it in Core 👍